### PR TITLE
New section Issuer state

### PR DESCRIPTION
**Pull request by adrifdez**
Resolves PID-569 PID-570 PID-572 PID-573 PID-574

To be able to publish a new state you should create a credential directly in the API or revoke it in the UI

You should have a positive balance in your wallet associated with your issuer

### CODE REVIEW

File: ui/src/domain/issuer-state.ts
Comment (cneuro) [inline ui/src/domain/issuer-state.ts:1]: This should be imported & exported in `domain/index.ts`.

File: ui/src/components/shared/CredentialRevokeModal.tsx [hunk 1]
```diff
@@ -29,18 +31,7 @@ export function CredentialRevokeModal({
         onClose();
         onRevoke();
```
Comment (cneuro) [inline ui/src/components/shared/CredentialRevokeModal.tsx:32]: This is missing (here or inside the callback itself) the additional notification that the credential has been revoked. See [comment](https://www.figma.com/file/BE3tUbvtSB2EqmJsybpeFC?node-id=5808:184164#417295665).

File: ui/src/components/shared/TableCard.tsx [hunk 2]
```diff
@@ -18,16 +18,16 @@ export function TableCard({
...
+  query?: string | null;
+  searchPlaceholder?: string;
   showDefaultContents: boolean;
   table: ReactNode;
   title: ReactNode;
 }) {
   return (
     <Card bodyStyle={{ padding: 0 }} title={title}>
-      {!showDefaultContents && (
+      {!showDefaultContents && onSearch && searchPlaceholder && query !== undefined && (
```
Comment (cneuro) [inline ui/src/components/shared/TableCard.tsx:30]: If `query` is defaulted to `null`, then checking for `query !== undefined` doesn't make sense?

File: ui/src/contexts/issuer-state.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        if (!isAbortedError(response.error)) {
+          void message.error(response.error.message);
+        }
+      }
+    },
+    [env]
+  );
+
+  const notifyChange = useCallback(() => {
+    void message.success({
```
Comment (cneuro) [inline ui/src/contexts/issuer-state.tsx:50]: I strongly believe this should be `message.info`. Let's [clarify](https://www.figma.com/file/BE3tUbvtSB2EqmJsybpeFC?node-id=5808:184164#417295665) this with Liudas.

File: ui/src/components/shared/CredentialRevokeModal.tsx (hunk 1 above)
Comment (adrifdez) [reply ui/src/components/shared/CredentialRevokeModal.tsx:32]: Sorry, but I don't get what it's missing, you are pointing to a line without changes and the notification is below that line

File: .env-ui.sample [hunk 4]
```diff
@@ -1,2 +1,3 @@
+BLOCK_EXPLORER_URL=https://mumbai.polygonscan.com
```
Comment (cneuro) [inline .env-ui.sample:1]: This and the local UI dev equivalent needs to be documented in the root README.

File: ui/src/adapters/api/issuer-state.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  }
+}
+
+export async function getStatus({
+  env,
+  signal,
+}: {
+  env: Env;
+  signal?: AbortSignal;
+}): Promise<APIResponse<boolean>> {
```
[the remark this answers is not in the document]
Comment (adrifdez) [reply ui/src/adapters/api/issuer-state.ts:75]: Could we call it `pendingActions` to be resued in the parser?

File: ui/src/adapters/api/issuer-state.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+      url: `${API_VERSION}/state/publish`,
+    });
+
+    return { data: true, isSuccessful: true };
+  } catch (error) {
+    return { error: buildAPIError(error), isSuccessful: false };
+  }
+}
+
+export async function retryState({ env }: { env: Env }): Promise<APIResponse<boolean>> {
```
[the remark this answers is not in the document]
Comment (adrifdez) [reply ui/src/adapters/api/issuer-state.ts:52]: retryPublishState

File: ui/src/components/shared/CredentialRevokeModal.tsx (hunk 1 above)
Comment (cneuro) [reply ui/src/components/shared/CredentialRevokeModal.tsx:32]: We need a `message.success` that displays "Credential successfully revoked" (or whatever the BE returns after revocation) in addition to the issuer state notification.

Comment (adrifdez) [reply ui/src/components/shared/CredentialRevokeModal.tsx:32]: I think that is more related to the decision that you'll  reach in figma, if we change it to an info message it'll has sense but if not I think that we can leave with just this one

Comment (cneuro) [reply ui/src/components/shared/CredentialRevokeModal.tsx:32]: Just for consistency though, we should always return the operation's message regardless of the issuer state notification.

File: ui/src/adapters/api/issuer-state.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (amonsosanz) [reply ui/src/adapters/api/issuer-state.ts:75]: ok let's keep the name `pendingActions`, I guess it makes sense if the API also uses this name.

File: ui/src/components/shared/CredentialRevokeModal.tsx [hunk 7]
```diff
@@ -3,8 +3,8 @@ import { useState } from "react";
 
 import { revokeCredential } from "src/adapters/api/credentials";
```
Comment (amonsosanz) [inline ui/src/components/shared/CredentialRevokeModal.tsx:4]: Can you please update the literal:

> Revoking of a credential must be accompanied by publishing of issuer state in order for  the action to be recorded

as:

> Revoking of a credential must be accompanied by publishing of issuer state in order for  the action to be effective

To be consistent with the delete connection msg?

File: ui/src/components/shared/CredentialRevokeModal.tsx [hunk 8]
```diff
@@ -3,7 +3,8 @@ import { useState } from "react";
```
Comment (amonsosanz) [inline ui/src/components/shared/CredentialRevokeModal.tsx:3]: Can you please update the literal:

> Revoking of a credential must be accompanied by publishing of issuer state in order for  the action to be recorded

as:

> Revoking of a credential must be accompanied by publishing of issuer state in order for  the action to be effective

To be consistent with the delete connection msg?

File: ui/src/contexts/IssuerState.tsx [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  }, [refreshStatus]);
+
+  const value = useMemo(() => {
+    return { notifyChange, refreshStatus, status };
+  }, [notifyChange, refreshStatus, status]);
+
+  return <IssuerStateContext.Provider value={value} {...props} />;
+}
+
+export function useStateContext() {
```
Comment (amonsosanz) [inline ui/src/contexts/IssuerState.tsx:77]: `useIssuerStateContext`

File: .env-ui.sample (hunk 4 above)
Comment (cneuro) [reply .env-ui.sample:1]: This also needs to be `ISSUER_UI_BLOCK_EXPLORER_URL` to remain consistent with the rest of the env vars.

File: ui/src/contexts/issuer-state.tsx (hunk 3 above)
Comment (cneuro) [reply ui/src/contexts/issuer-state.tsx]: Hi @Figmintojas could you quickly weigh in here? Basically, instead of this:

Screenshot from 2023-04-20 12-19-48

I am advocating to [keep messaging consistent](https://www.figma.com/file/BE3tUbvtSB2EqmJsybpeFC/Polygon-ID-Platform?node-id=5806-174811&t=92KxN4W0RYdvJLdK-0) by always accompany it with the operational feedback and changing the icon to represent the message type. Like this:

Screenshot from 2023-04-20 12-21-29

File: ui/src/components/issuer-state/IssuerState.tsx [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+        clearInterval(checkUnpublished);
+      }
+    }, POLLING_INTERVAL);
+
+    return () => clearInterval(checkUnpublished);
+  }, [fetchTransactions, transactionsList]);
+
+  return (
+    <SiderLayoutContent
+      description="Issuing merkle tree type credentials and revoking credentials require an additional step, know as publishing issuer state."
```
Comment (cneuro) [inline ui/src/components/issuer-state/IssuerState.tsx:194]: "Merkle" should be [capitalized](https://en.wikipedia.org/wiki/Merkle_tree).

File: ui/src/contexts/issuer-state.tsx (hunk 3 above)
Comment (amonsosanz) [reply ui/src/contexts/issuer-state.tsx]: I think we can go ahead with this. It's pretty obvious you're right @cneuro. The first msg is info, the second is a success op confirmation.

File: README.md [hunk 11]
```diff
@@ -125,12 +126,12 @@ Completing either option of the [installation](#installation) process yields the
...
-    - `VITE_API_PASSWORD -> ISSUER_API_UI_AUTH_PASSWORD`
-    - `VITE_ISSUER_DID -> ISSUER_API_UI_ISSUER_DID`
-    - `VITE_ISSUER_NAME -> ISSUER_API_UI_ISSUER_NAME`
-    - `VITE_ISSUER_LOGO -> ISSUER_API_UI_ISSUER_LOGO`
+   - `VITE_API_URL -> ISSUER_API_UI_SERVER_URL`
+   - `VITE_API_USERNAME -> ISSUER_API_UI_AUTH_USER`
+   - `VITE_API_PASSWORD -> ISSUER_API_UI_AUTH_PASSWORD`
+   - `VITE_ISSUER_DID -> ISSUER_API_UI_ISSUER_DID`
+   - `VITE_ISSUER_NAME -> ISSUER_API_UI_ISSUER_NAME`
+   - `VITE_ISSUER_LOGO -> ISSUER_API_UI_ISSUER_LOGO`
```
Comment (amonsosanz) [inline README.md:134]: isn't `VITE_BLOCK_EXPLORER_URL` missing here?

File: ui/src/components/issuer-state/IssuerState.tsx [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+  const [isPublishing, setIsPublishing] = useState<boolean>(false);
+  const [isRetrying, setIsRetrying] = useState<boolean>(false);
+  const [transactions, setTransactions] = useState<AsyncTask<Transaction[], APIError>>({
+    status: "pending",
+  });
+
+  const transactionsList = useMemo(
+    () => (isAsyncTaskDataAvailable(transactions) ? transactions.data : []),
+    [transactions]
+  );
```
Comment (amonsosanz) [inline ui/src/components/issuer-state/IssuerState.tsx:47]: Why did you need to use `useMemo` in this case?

File: ui/src/contexts/IssuerState.tsx [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+      } else {
+        if (!isAbortedError(response.error)) {
+          void message.error(response.error.message);
+        }
+      }
+    },
+    [env]
+  );
+
+  const notifyChange = useCallback(() => {
```
Comment (amonsosanz) [inline ui/src/contexts/IssuerState.tsx:49]: Currently, this does not inform the user about any type of pending to-be-published change. It's very specific to revocation changes. If this is expected to remain specific, I'd rename it as `notifyPendingRevocationChange`.

I suspect though that this msg will be also required to be shown when we issue new direct credentials. In such a case, I'd keep the name as it's now but require the msg `Revocation requires issuer state to be published` to be passed as a param.

File: ui/src/adapters/api/issuer-state.ts [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    const response = await axios({
+      baseURL: env.api.url,
+      headers: {
+        Authorization: buildAuthorizationHeader(env),
+      },
+      method: "GET",
+      signal,
+      url: `${API_VERSION}/state/transactions`,
+    });
+    const { data } = resultOKTransactionsParser.parse(response);
```
Comment (amonsosanz) [inline ui/src/adapters/api/issuer-state.ts:118]: I'd sort the tx's so the newer ones are at the top, as we do with the schemas.

File: ui/src/components/issuer-state/IssuerState.tsx [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+      }
+
+      void refreshStatus();
+      void fetchTransactions();
+
+      setIsPublishing(false);
+    });
+  };
+
+  const retry = () => {
```
Comment (cneuro) [inline ui/src/components/issuer-state/IssuerState.tsx:69]: This is almost identical to `publish` ... can we DRY this up? Maybe at least with the success message?

File: ui/src/contexts/IssuerState.tsx (hunk 13 above)
Comment (adrifdez) [reply ui/src/contexts/IssuerState.tsx:49]: I'd remove this and reuse `refreshStatus` then

**Review by cneuro — APPROVED** [review]

**Review by amonsosanz — APPROVED** [review]
